include/material/modified_cam_clay.tcc

```diff
@@ -0,0 +1,722 @@
+    (*state_vars).at("m_theta") = m_theta_trial;
+    // Compute dF / dmul
+    this->compute_df_dmul(state_vars, &df_dmul);
+    // Update consistency parameter
+    (*state_vars).at("delta_phi") -= ((*state_vars).at("f_function") / df_dmul);
+    // Initialise G and dG / dpc
+    double g_function = 0;
+    double dg_dpc = 0;
+    // Compute G and dG / dpc
+    this->compute_dg_dpc(state_vars, pc_n, p_trial, &g_function, &dg_dpc);
+    // Subiteraction for preconsolidation pressure
+    while (fabs(g_function) > Gtolerance && counter_g < substep) {
+      // Update preconsolidation pressure
+      (*state_vars).at("pc") -= g_function / dg_dpc;
+      // Update G and dG / dpc
+      this->compute_dg_dpc(state_vars, pc_n, p_trial, &g_function, &dg_dpc);
+      // Counter subiteration step
+      ++counter_g;
+    }
+    // Update mean pressure p
+    (*state_vars).at("p") =
+        (p_trial + (*state_vars).at("bulk_modulus") *
+                       (*state_vars).at("delta_phi") * (*state_vars).at("pc")) /
+        (1 +
+         2 * (*state_vars).at("bulk_modulus") * (*state_vars).at("delta_phi"));
+    // Update deviatoric stress q
+    // Equation(3.10b)
+    (*state_vars).at("q") =
+        q_trial / (1 + 6 * (*state_vars).at("shear_modulus") *
+                           (*state_vars).at("delta_phi") /
+                           pow((*state_vars).at("m_theta"), 2));
+    // Compute incremental plastic volumetic strain
+    // Equation(2.8)
+    (*state_vars).at("dpvstrain") =
+        (*state_vars).at("delta_phi") *
+        (2 * (*state_vars).at("p") - (*state_vars).at("pc") -
+         (*state_vars).at("pcd"));
+    // Compute plastic deviatoric strain
+    (*state_vars).at("dpdstrain") =
+        (*state_vars).at("delta_phi") *
+        (sqrt(6) * (*state_vars).at("q") / pow((*state_vars).at("m_theta"), 2));
+    // Update bonding parameters
+    if (bonding_) this->compute_bonding_parameters(chi_n, state_vars);
+    // Compute subloading parameters
+    if (subloading_)
+      this->compute_subloading_parameters(subloading_r, state_vars);
+    // Compute three invariants parameters
+    if (three_invariants_) {
+      // Update stress
+      // Type-1 Equation(3.16)
+      updated_stress = (*state_vars).at("q") * n_trial;
+      for (int i = 0; i < 3; ++i) updated_stress(i) -= (*state_vars).at("p");
+      // Compute stress invariants
+      this->compute_stress_invariants(updated_stress, n_trial, state_vars);
+      // Update Mtheta
+      (*state_vars).at("m_theta") =
+          m_ - pow(m_, 2) / (3 + m_) * cos(1.5 * (*state_vars).at("theta"));
+    }
+    // Update yield function
+    yield_type = this->compute_yield_state(state_vars);
+    // Counter iteration step
+    ++counter_f;
+  }
+
+  // Update stress
+  updated_stress = (*state_vars).at("q") * n_trial;
+  for (int i = 0; i < 3; ++i) updated_stress(i) -= (*state_vars).at("p");
+
+  // Update void_ratio
+  (*state_vars).at("void_ratio") +=
```

Comment:
Just a small note that we haven't figure out the connection between void_ratio and porosity @tianchiTJ @srhgk2. Generalizing them will be good.